IMU: parameterize IMU integration time (IMU_INTEG_RATE)

dagar opened the pull request:
This effectively controls the `vehicle_attitude` publication rate and subsequently how fast the attitude controllers run. It's not something I expect regular users to to change, but this is better than having a magic hard coded number.

We might want to consider lowering this by default for FW.

dagar commented:
Historically the IMU integration time was  4000 us (250 Hz) and only very recently changed to 2500 us (400 Hz). We could also consider if 5000 us (200 Hz) is a better default.

dagar commented:
Changing the integration from 2500 us -> 5000 us saves about 7% cpu on F4 boards.

dagar commented:
Yet another factor to consider for the default is the impact on default logging rate and ekf2 replay.

On a typical simple multicopter configuration the nominal logging rate changes from ~ 47 KiB/s (IMU integration 2500 us) down to ~32 KiB/s (IMU integration 5000 us).

MaEtUgR commented on src/lib/drivers/gyroscope/PX4Gyroscope.cpp:
```diff
@@ -122,11 +125,19 @@ void PX4Gyroscope::set_device_type(uint8_t devtype)
+
+	if (imu_integration_time_us != _param_imu_integ_time.get()) {
+		_param_imu_integ_time.set(imu_integration_time_us);
+		_param_imu_integ_time.commit_no_notification();
```
Isn't it a problem that all Gyroscope and Accelerometer drivers would try to correct the out of range parameter more or less at the same time?

MaEtUgR replied:
I would at least define the [1000, 20000] interval shared to make sure it's always in sync.

dagar replied:
Yes I didn't like that part either. Very soon (early in the v1.12 release cycle) the plan is to gut all of this code from PX4Accelerometer/PX4Gyroscope and move it "downstream" to sensors/vehicle_imu where we can effectively synchronize accel + gyro.

MaEtUgR commented:
One more question I forgot to ask in the review: Is there a specific reason why the new parameter is an interval while all others I know of are rates in Hz? Just thinking about the user perspective.

dagar commented:
> One more question I forgot to ask in the review: Is there a specific reason why the new parameter is an interval while all others I know of are rates in Hz? Just thinking about the user perspective.

Not really, I was thinking about it from the perspective of what was already hard coded and aligning with the current ecl/EKF filter update expressed as an interval (10 ms). We reset the integrator and publish the integrated value every ____ microseconds.

dagar commented:
Default changed to 5000 us.

dagar commented:
SITL appears to be broken with 5000 us IMU integration.

julianoes commented:
@dagar we might have to adapt the gazebo side too:

dagar commented:
Testing with sitl_gazebo and jmavsim updated as well.
 - https://github.com/PX4/sitl_gazebo/pull/474
 - https://github.com/PX4/jMAVSim/pull/118

dagar commented:
This was tested by @mhkabir in https://github.com/PX4/Firmware/issues/14735.

dagar commented on src/modules/sensors/vehicle_imu/imu_parameters.c:
```diff
@@ -0,0 +1,46 @@
+ ****************************************************************************/
+
+/**
+* IMU integration period.
```
```suggestion
* IMU integration time.
```

dagar commented:
CI SITL MC_avoidance and MC_safe_landing still failing due to the sitl_gazebo version.

dagar commented:
I've made 2 minor changes to the PR.
1. renamed IMU_INTEG_TIME -> IMU_INTEG_RATE (I think it's slightly easier for most people to think in terms of rate)
2. Dropped the sitl_gazebo and jmavsim changes for now until we can resolve the performance problems https://github.com/PX4/sitl_gazebo/pull/474.
